addons/mrp/report/mrp_report_bom_structure.py

```diff
@@ -122,6 +123,65 @@ def _get_report_data(self, bom_id, searchQty=0, searchVariant=False):
+        for line in lines:
+            product = line.product_id
+            quantities_info = self._get_quantities_info(product, line.product_uom_id, parent_bom, product_info)
+            stock_loc = quantities_info['stock_loc']
+            product_info[product.id]['consumptions'][stock_loc] += line_quantities.get(line.id, 0.0)
+            product_quantities_info[product.id][line.id] = product_info[product.id]['consumptions'][stock_loc]
+            if product.detailed_type != 'product':
+                # Use date.min as a sentinel value for _get_stock_availability
+                closest_forecasted[product.id][line.id] = date.min
+            elif stock_loc != 'in_stock':
+                closest_forecasted[product.id][line.id] = date.max
+            elif float_compare(product_info[product.id]['consumptions'][stock_loc], quantities_info['free_qty'], precision_rounding=product.uom_id.rounding) <= 0:
+                closest_forecasted[product.id][line.id] = date.min
+            else:
+                remaining_products.append(product.id)
+                closest_forecasted[product.id][line.id] = None
+        date_today = self.env.context.get('from_date', fields.date.today())
+        domain = [('state', '=', 'forecast'), ('date', '>=', date_today), ('product_id', 'in', list(set(remaining_products)))]
+        if self.env.context.get('warehouse'):
+            domain.append(('warehouse_id', '=', self.env.context.get('warehouse')))
+        if remaining_products:
+            res = self.env['report.stock.quantity']._read_group(
+                domain,
+                ['min_date:min(date)', 'product_id', 'product_qty'],
+                ['product_id', 'product_qty'],
+                orderby='product_id asc, min_date asc', lazy=False
+            )
+            available_quantities = defaultdict(list)
+            for group in res:
+                product_id = group['product_id'][0]
+                available_quantities[product_id].append([group['min_date'], group['product_qty']])
+            for product_id in remaining_products.copy():
+                # Find the first empty line_id for the given product_id.
+                line_id = next(filter(lambda k: not closest_forecasted[product_id][k], closest_forecasted[product_id].keys()), None)
+                # Find the first available quantity for the given product and update closest_forecasted
+                for min_date, product_qty in available_quantities[product_id]:
+                    if product_qty >= product_quantities_info[product_id][line_id]:
+                        closest_forecasted[product_id][line_id] = min_date
+                        remaining_products.remove(product_id)
+                        break
+            for product_id in remaining_products:
+                for key, val in closest_forecasted[product_id].items():
+                    if not val:
+                        closest_forecasted[product_id][key] = date.max
+                        break
```

Comment:
From what I understand, for each product 'line' that doesn't have a date set (i.e. `available` or `not available`), we compare it to the forecast results to see if a date can be found where the demand is met.
If no date was found for the product 'line', then the default `date.max` will be set. 

Couldn't it be simplified to this?
```suggestion
            for product_id in remaining_products:
                # Find the first empty line_id for the given product_id.
                line_id = next(filter(lambda k: not closest_forecasted[product_id][k], closest_forecasted[product_id].keys()), None)
                # Find the first available quantity for the given product and update closest_forecasted
                for min_date, product_qty in available_quantities[product_id]:
                    if product_qty >= product_quantities_info[product_id][line_id]:
                        closest_forecasted[product_id][line_id] = min_date
                        break
                if not closest_forecasted[product_id][line_id]:
                    closest_forecasted[product_id][line_id] = date.max
```
I understand that doing so will lead to a different evaluation of `product_quantities_info[product_id][line_id]` (since the `line_id` will be the next in line), but since the quantities are additives, if no dates were found for the first line, I don't see how one could be found for the next one. :thinking:

Reply:
@clesgow Hello, I was on vacation so that's why I couldn't apply the changes earlier. I think you're right about that part, I lost myself a bit in complexity here. I made the changes you suggested.